CMakeLists.txt

```diff
@@ -1,504 +1,2025 @@
+  target_compile_options (ripple_boost INTERFACE --system-header-prefix="boost/")
+else ()
+  target_include_directories (ripple_boost SYSTEM INTERFACE ${Boost_INCLUDE_DIRS})
 endif()
 
-add_with_props(rippled_src_all src/ripple/unity/secp256k1.cpp
-  -I"${CMAKE_SOURCE_DIR}/"src/secp256k1
-  ${no_unused_w}
-  )
-
-foreach(cursrc
-    src/ripple/unity/beast.cpp
-    src/ripple/unity/lz4.c
-    src/ripple/unity/protobuf.cpp
-    src/ripple/unity/ripple.proto.cpp)
-
-  add_with_props(rippled_src_all ${cursrc}
-    ${rocks_db_system_header}
+target_link_libraries (ripple_boost
+  INTERFACE
+    Boost::boost
+    Boost::chrono
+    Boost::coroutine
+    Boost::date_time
+    Boost::filesystem
+    Boost::program_options
+    Boost::regex
+    Boost::serialization
+    Boost::system
+    Boost::thread)
+
+#[===================================================================[
+   NIH dep: openssl
+#]===================================================================]
+
+if (APPLE AND NOT DEFINED ENV{OPENSSL_ROOT_DIR})
+  find_program (HOMEBREW brew)
+  if (NOT HOMEBREW STREQUAL "HOMEBREW-NOTFOUND")
+    execute_process (COMMAND ${HOMEBREW} --prefix openssl
+      OUTPUT_VARIABLE OPENSSL_ROOT_DIR
+      OUTPUT_STRIP_TRAILING_WHITESPACE)
+  endif ()
+endif ()
+
+if ((NOT DEFINED OPENSSL_ROOT) AND (DEFINED ENV{OPENSSL_ROOT}))
+  set (OPENSSL_ROOT $ENV{OPENSSL_ROOT})
+endif ()
+file (TO_CMAKE_PATH "${OPENSSL_ROOT}" OPENSSL_ROOT)
+
+if (static OR APPLE OR MSVC)
+  set (OPENSSL_USE_STATIC_LIBS ON)
+endif ()
+set (OPENSSL_MSVC_STATIC_RT ON)
+set (_ssl_min_ver 1.0.2)
+# HACK for travis
+if ("$ENV{CI}" STREQUAL "true" AND "$ENV{TRAVIS}" STREQUAL "true")
+  set (_ssl_min_ver 1.0.1)
+endif ()
+find_package (OpenSSL ${_ssl_min_ver} REQUIRED)
+target_link_libraries (ripple_libs
+  INTERFACE
+    OpenSSL::SSL
+    OpenSSL::Crypto)
+# disable SSLv2...this can also be done when building/configuring OpenSSL
+set_target_properties(OpenSSL::SSL PROPERTIES
+    INTERFACE_COMPILE_DEFINITIONS OPENSSL_NO_SSL2)
+#[=========================================================[
+   https://gitlab.kitware.com/cmake/cmake/issues/16885
+   depending on how openssl is built, it might depend
+   on zlib. In fact, the openssl find package should
+   figure this out for us, but it does not currently...
+   so let's add zlib ourselves to the lib list
+   TODO: investigate linking to static zlib for static
+   build option
+#]=========================================================]
+find_package (ZLIB)
+if (TARGET ZLIB::ZLIB)
+  #target_link_libraries (OpenSSL::Crypto INTERFACE ZLIB::ZLIB)
+  set_target_properties(OpenSSL::Crypto PROPERTIES
+    INTERFACE_LINK_LIBRARIES ZLIB::ZLIB)
+endif ()
+
+#[===================================================================[
+   NIH dep: secp256k1
+#]===================================================================]
+
+add_library (secp256k1
+  src/secp256k1/src/secp256k1.c)
```

Comment:
This changes the way rippled is built from a large set of object files all linked into a single executable to a set of libraries that are linked together. I don't object to that. However, these libraries are built as shared libraries (for non-static configurations). Do we want to force these libraries to be statically linked? I can't think of any benefits to a shared library for these sub-libraries, are there any?

Reply:
I'm not sure what problem making these libraries dynamic solves. I think of these libraries as a "convenient way to set compile options on a set of files" and not really as stand alone libraries. I'd rather keep them internal to rippled.

However, I don't object too strongly to allowing dynamic linking, but lets think about two things before we close this issue:

1) Do we have to worry about version mismatches - i.e. running a new version of rippled with an old version of the dynamic libraries? Should we add a version number to the libraries?

2) Do we have to worry about name conflicts? We have libraries called `libbeast_legacy.so  libcorepp.so  libed25519-donna.so  libsecp256k1.so`. Some of those sound generic enough that another library might have the same name. Maybe add a `rippled_` prefix? Also, we should rename `libcorepp.so` at any rate.

If the above two things can be addressed then I'm OK with allowing dynamic linking of these new libraries.

Reply:
I try to stay out of the way when folks are working on tools, since I really appreciate not having to do it myself.  But I'm going to side with @seelabs on this one.  I have historically, at other work places, been a strong proponent for static linking as much as possible.  It moves complexity (and potential for error) to link time rather than at (the beginning of) runtime.  That gives a developer a better chance at catching errors and reduces the number of errors that users potentially have to deal with.

So, unless there are strong reasons to switch to dynamic linking, I'd prefer that the libraries be statically linked.  Just my take...

Reply:
## linking ##

@scott.determan which libraries specifically are you referring to? corepp and
beast_legacy are most definitely NOT just a set of compile options - they
contain executable code that is reusable for certain applications (e.g.
signing). Any of the `INTERFACE` libraries here are indeed just bundles
of settings, but then those libraries don't have any installed artifacts other
than being part of the config files that are installed by our build and used by
clients of the libs. I should say, also, that I consider beast_legacy and
corepp really one library conceptually in that they serve a singular purpose
of exposing some of the core rippled functionality to client applications. The
fact that there are two libraries is just a reflection of the fact that
beast_legacy is code that we'd actually like to delete and/or replace (there
was a specific task to "break out" the old beast code to make it more obvious
what old code we still rely on and need to look at replacing. This is more
organizational than logical at this point).

As far as keeping things internal to rippled, I was under the impression that
we want to expose some functionality in the form of library. ripple-libpp was
one take on this. corepp (to be better named) is a second take on the same
concept and one that is properly installed/configured and can be rightly
included in a -devel package, for instance. Do you object fundamentally to the
concept of ripple-libpp?

*If* you agree that we want to provide a library for some client applications
to use, then comes the question of how to install and link to it. I'm saying
the linking type (dynamic vs. static) can be left to the user. Most libraries
out there leave this to the user. Even OpenSSL allows dynamic linking if you
want to, although they tend to make static the default. I'm saying we can do
the same - default to static, but allow dynamic if that's what you want. If
I were charged with building a -devel RPM, I'd probably just install a static
lib, but I could be convinced otherwise. Maybe we leave the dynamic lib option
to "advanced" users, which means you'd have to build from source.

There is also the somewhat related concept of how to link to third-party (NIH)
dependencies. Up until now, we've taken a bizarre hybrid approach of direct
source inclusion for some libraries (sqlite, secp256k, protobuf on windows,
etc.) and standard configure time discovery of system libs (openssl and
boost). The direct inclusion deps are implicitly statically linked (really,
object file linked - but it's all very non-transparent what's going on when
you do that). I find this duplicity somewhat confusing personally, and I've
attempted in this change to at least treat all NIH, as much as possible, the same.
Even if we subtree'd the dependency, I wanted to build it separately
becase *it is* fundamentally a third party library. It's not our code and yet
we just dropped it right into our exe build previously.

If we want to make the decision that all NIH deps are linked statically, that's
certainly one approach, but we would be rolling back
https://github.com/ripple/rippled/pull/2291, for instance. Or perhaps we *only*
want to allow optional dynamic linking to certain third party libs? If so,
which ones? Just boost? How about OpenSSL? If you force NIH to be static,
but allow corepp to be built dynamically, then you will have a dynamic lib
that depends on a static lib, which I always think is a little weird, but
still feasible. Again, if we want to make the sweeping decision that ALL
libs involved in rippled (both NIH and our libs) are static, then it
simplifies things somewhat but ties users' hands with respect to
link-type.

## versioning ##

If you are going to have multiple versions of rippled + libs installed
on your system, then you'd need to manage that like you
would any other libs. How would you handle multiple versions of OpenSSL or
boost? You'd probably install into separate prefixes. Same thing applies here -
for example:

```
/opt/rippled/1.0/
  .../bin
  .../lib
  .../include
/opt/rippled/1.1/
  .../bin
  .../lib
  .../include
```

Remember that cmake links using rpath, so the binary should have a pretty
strong link to the library it was installed with.

CMake supports [SOVERSION](https://cmake.org/cmake/help/latest/prop_tgt/SOVERSION.html),
so if we think that's the best practice, we can certainly do that for our
library. This is effectively version 1 of our library and we've not made any
attempt to version our API up to this point. If we want to start lib/API
versioning now, then I guess SOVERSION is a place to start.

One other note, the cmake config package that is created for our library supports
version checking and compatibility and I've chosen the "SameMajorVersion" compat check
for now, but we can certainly change that if we want:
https://github.com/ripple/rippled/blob/735c8d0486b0dc69683a123e6c4f520c505a4fcc/CMakeLists.txt#L1948-L1951
This version checking covers you at build time if you are building a client
application of some kind using the packages/libs we install.

## name conflicts ##

I think the NIH libs are the most likely to potentially conflict (sqlite, for
example), but that's mostly a reflection of our decision to *build* those
libs ourselves instead of relying on installed/system packages. Given that
we *do* have these private builds of several NIH libs, I'd be inclined to
advise against installing rippled in common paths like `/usr` or `/usr/local`
because of the potential to collide with system libs. Again, rpath will
help us here, although universally switching to static linking also obviates
this issue. I think in its present form, rippled should be generally be
installed using it's own prefix like `/opt/rippled`, for example.

## naming ##

Now the hardest part of this entire project: naming the library. Here are some
ideas, but I'm not married to anything (please suggest something better...):

libripplepp
libledgersign
libxrpproto

If you want a different name for beast_legacy, just say so - I consider it less
public facing so I'm not quite as concerned about that one.

Reply:
I had assumed that all these libraries were internal to rippled and were not meant to be used by themselves. If `corepp` is meant to be used stand alone then I agree that the user should be in charge of the link type and we should provide both static and dynamic options.

I agree is conceptually simpler to treat everything the same and in a dynamic build not force some libraries to be static. I do think we want to static link by default, but given `corepp` is meant to be used stand alone what you have here is best.

I don't know what the best way to handle versioning is, but since these are handled through rpath and presumably won't be on LD_LIBRARY_PATH I guess it doesn't matter.

For names:
I'd use the same prefix for all of them, maybe `xrpl` for xrp ledger? I wouldn't use `pp`, it's too cryptic. Maybe `xrpl_core`?

Reply:
ah, sorry if that wasn't clear. Yes, corepp is intended to supplant https://github.com/ripple/ripple-libpp.  I like the `xrpl` prefix, and I want to see if I can just roll `beast_legacy` into an object lib so we don't have to deal with an extra artifact.